### PR TITLE
Resolves #25 - An authenticated user can post a question

### LINKED ISSUE #25: An authenticated user can post a question
**Scenario**

Given I'm an authenticated user of any type
When I send a POST request to /discussions/questions with the required input
I expect that question to be stored

**Description**

One of the main objectives of the Laravel-Portugal community is to help our fellow developers overcome their difficulties and have their doubts answered by other developers. This type of interaction has been supported by the normal operations of our Discord server, which has specific channels where registered members can post questions and start discussing them.

However, when many developers start to submit any questions, at the same time, the conversation gets to an unbearable and unmanageable point that to avoid such a mess, many people start to move the discussion to private chats and won't benefit other devs with the same doubts because it's unsearchable.

Since we're developing a new portal, and we've planned to introduce [users' accounts](https://github.com/laravel-portugal/api/issues/22), it's time to start adding the possibility to concentrate all questions, and the knowledge shared while resolving them, in a publicly available place. It's also a good way to focus the Discord server to only be a place of synchronous, real-time, conversations.

**Database and schema information**

A new `questions` table should be created with the following structure:
- `id`: PK.
- `author_id`: FK, related to the user who created the question. Must be required and indexed.
- `title`: the title of the question. Must be required and indexed.
- `slug`: the URL friendly identifier, automatically generated based on the title.
- `description`: the question itself. Must be required.
- `created_at`: timestamp with timezone. Must be indexed and default to the date the resource is created.
- `updated_at`: timestamp with timezone. Must be indexed.
- `deleted_at`: timestamp with timezone. Must be indexed and default to null. This indicates if the resource was deleted.
- `resolved_at`: timestamp with timezone. Must be indexed and default to null. This indicates that the question as resolved.

**Additional requirements**

This builds on the features of #22, which introduces users' accounts.

### PR DESCRIPTION
**Pull request by tiagof**

### CODE REVIEW

File: domains/Accounts/Controllers/AccountsStoreController.php [hunk 1]
```diff
@@ -11,13 +11,6 @@
 
 class AccountsStoreController extends Controller
 {
-    private User $user;
-
-    public function __construct(User $users)
-    {
-        $this->users = $users;
-    }
-
```
Comment (josepostiga) [inline domains/Accounts/Controllers/AccountsStoreController.php:20]: I think I understand why this was removed. But I think we should compose our controllers and inject any dependency and reference them when needed. 

But we should do it on another PR, to not pollute this one with code that has little to do with it. Can you revert this change, for now?

File: domains/Discussions/Controllers/QuestionsStoreController.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+class QuestionsStoreController extends Controller
+{
+    public function __invoke(Request $request): Response
+    {
+        $this->validate($request, [
+            'title' => ['required', 'string', 'max:255'],
+            'description' => ['required', 'string'],
+        ]);
+
+        (new Question())
```
Comment (josepostiga) [inline domains/Discussions/Controllers/QuestionsStoreController.php:20]: Let's inject this Question class as a dependency from the constructor and reference it here, like it's done in the `LinksStoreController` for example.

File: domains/Discussions/Controllers/QuestionsStoreController.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+            ->forceFill([
+                'author_id' => Auth::id(),
```
Comment (josepostiga) [inline domains/Discussions/Controllers/QuestionsStoreController.php:22]: We should use the `$request` class to et the authenticated user's id.

File: domains/Discussions/Database/Migrations/2020_10_12_104113_create_questions_table.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('questions', function (Blueprint $table) {
+            $table->id();
+            $table->foreignIdFor(User::class, 'author_id');
```
Comment (josepostiga) [inline domains/Discussions/Database/Migrations/2020_10_12_104113_create_questions_table.php:19]: I believe creating foreign keys doesn't implicitly create an index. We need to create one explicitly.

File: domains/Discussions/Database/Migrations/2020_10_12_104113_create_questions_table.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    /**
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
+    {
+        Schema::dropIfExists('questions');
+    }
```
Comment (josepostiga) [inline domains/Discussions/Database/Migrations/2020_10_12_104113_create_questions_table.php:37]: Nit: we don't use the down method.

File: domains/Discussions/Database/Migrations/2020_10_12_104113_create_questions_table.php [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+use Illuminate\Support\Facades\Schema;
+
+class CreateQuestionsTable extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
```
Comment (josepostiga) [inline domains/Discussions/Database/Migrations/2020_10_12_104113_create_questions_table.php:15]: Nit: the dockblocks are unnecessary because they don't give any more useful information.

Also, we should add the return type hints.

File: domains/Discussions/Models/Question.php [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+use Illuminate\Database\Eloquent\Relations\BelongsTo;
+use Illuminate\Database\Eloquent\SoftDeletes;
+
+class Question extends Model
+{
+    use SoftDeletes;
+
+    public function author(): BelongsTo
+    {
+        return $this->belongsTo(User::class);
```
Comment (josepostiga) [inline domains/Discussions/Models/Question.php:16]: We should take into consideration that a User resource can be soft-deleted. However, we might not want to delete resources that that User created. If we don't define this relationship with the `withTrashed()` scope, we won't have access to the soft-deleted User information.

I think that, by default, we should load all relationship information.

File: domains/Discussions/Observers/QuestionObserver.php [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace Domains\Discussions\Observers;
+
+
+use Domains\Discussions\Models\Question;
+use Illuminate\Support\Str;
+
+class QuestionObserver
+{
+    public function creating(Question $question)
```
Comment (josepostiga) [inline domains/Discussions/Observers/QuestionObserver.php:12]: Don't forget the return type hints.

File: domains/Discussions/Tests/Feature/QuestionsStoreTest.php [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    /** @test */
+    public function it_stores_questions(): void
+    {
+        $payload = [
+            'title' => $this->faker->title,
+            'description' => $this->faker->paragraph,
+        ];
+
+        $this->actingAs($this->user);
+        $response = $this->call('POST', '/questions', $payload);
```
Comment (josepostiga) [inline domains/Discussions/Tests/Feature/QuestionsStoreTest.php]: Nit: we can chain these two calls together.

File: domains/Discussions/Tests/Feature/QuestionsStoreTest.php [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+            'title' => $payload['title'],
+            'description' => $payload['description'],
+        ]);
+    }
+
+    /** @test */
+    public function it_calculates_a_slug(): void
+    {
+        self::assertEquals(Str::slug($this->question->title), $this->question->slug);
+    }
```
Comment (josepostiga) [inline domains/Discussions/Tests/Feature/QuestionsStoreTest.php:56]: We could add this assertion in `it_stores_questions()` test case. It'll also remove the need to pre-create a Question resource for this, which I find it weird to happen.

File: domains/Discussions/Tests/Feature/QuestionsStoreTest.php [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        $this->actingAs($this->user);
+
+        $response = $this->call('POST', '/questions', [
+            'title' => $this->question->title, // Use same 'title' as the Question created in setUp()
+            'description' => $this->question->description,
+        ]);
+
+        self::assertTrue($response->isEmpty());
+        self::assertEquals(2, Question::query()->count());
+    }
```
Comment (josepostiga) [inline domains/Discussions/Tests/Feature/QuestionsStoreTest.php]: I don't think this is a test case that we should be worrying about. We really don't care about duplicating slugs. The only thing we care about is that each question gets its own unique internal ID, and that's the default behavior of the persistent layer

**Review by josepostiga — CHANGES_REQUESTED** [review]
As usual, this is looking really good. :+1: Thanks for the effort, and for the patience while I found the time to review this.

I've requested a few changes, however, that I think makes sense to apply here. Let's iterate further on this implementation.

File: domains/Accounts/Controllers/AccountsStoreController.php (hunk 1 above)
Comment (tiagof) [reply domains/Accounts/Controllers/AccountsStoreController.php:20]: sure

File: domains/Discussions/Controllers/QuestionsStoreController.php (hunk 2 above)
Comment (tiagof) [reply domains/Discussions/Controllers/QuestionsStoreController.php:20]: done

File: domains/Discussions/Controllers/QuestionsStoreController.php (hunk 3 above)
Comment (tiagof) [reply domains/Discussions/Controllers/QuestionsStoreController.php:22]: done

File: domains/Discussions/Database/Migrations/2020_10_12_104113_create_questions_table.php (hunk 4 above)
Comment (tiagof) [reply domains/Discussions/Database/Migrations/2020_10_12_104113_create_questions_table.php:19]: It does in MySQL but not in others, so yes, let's explicitly add it.

File: domains/Discussions/Models/Question.php (hunk 7 above)
Comment (tiagof) [reply domains/Discussions/Models/Question.php:16]: makes sense !

File: domains/Discussions/Tests/Feature/QuestionsStoreTest.php (hunk 10 above)
Comment (tiagof) [reply domains/Discussions/Tests/Feature/QuestionsStoreTest.php:56]: Ok

File: domains/Discussions/Tests/Feature/QuestionsStoreTest.php (hunk 11 above)
Comment (tiagof) [reply domains/Discussions/Tests/Feature/QuestionsStoreTest.php]: Ok, i'll remove it then.

File: domains/Discussions/Tests/Feature/QuestionsStoreTest.php [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    private User $user;
+    private Question $question;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->faker = Factory::create();
+        $this->user = UserFactory::new()->create();
+        $this->question = QuestionFactory::new(['author_id' => $this->user->id])->create();
```
Comment (josepostiga) [inline domains/Discussions/Tests/Feature/QuestionsStoreTest.php:30]: It looks like this is no longer needed.

File: domains/Discussions/Tests/Feature/QuestionsStoreTest.php [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    {
+        $payload = [
+            'title' => $this->faker->title,
+            'description' => $this->faker->paragraph,
+        ];
+
+        $response = $this->actingAs($this->user)
+            ->call('POST', route('discussions.questions.store'), $payload);
+
+        self::assertTrue($response->isEmpty());
```
Comment (josepostiga) [inline domains/Discussions/Tests/Feature/QuestionsStoreTest.php:44]: Care to check the response status code is the correct one? It should be 204.

**Review by josepostiga — CHANGES_REQUESTED** [review]
Nice job, here, once again. Just left a couple of notes and we're ready to merge.

Feel free to merge after doing the necessary changes and the test suite passes :+1: 

Also, don't forget to update the changelog file.

File: domains/Discussions/Tests/Feature/QuestionsStoreTest.php (hunk 12 above)
Comment (tiagof) [reply domains/Discussions/Tests/Feature/QuestionsStoreTest.php:30]: Done

File: domains/Discussions/Tests/Feature/QuestionsStoreTest.php (hunk 13 above)
Comment (tiagof) [reply domains/Discussions/Tests/Feature/QuestionsStoreTest.php:44]: done
